contracts/andromeda_vault/src/contract.rs

```diff
@@ -0,0 +1,426 @@
+    parse_message, require,
+    withdraw::{Withdrawal, WithdrawalType},
+};
+use cosmwasm_std::{
+    coin, entry_point, to_binary, BankMsg, Binary, Coin, ContractResult, CosmosMsg, Deps, DepsMut,
+    Empty, Env, MessageInfo, Order, QueryRequest, ReplyOn, Response, SubMsg, SystemResult, Uint128,
+    WasmMsg, WasmQuery,
+};
+use cw2::{get_contract_version, set_contract_version};
+
+// version info for migration info
+const CONTRACT_NAME: &str = "crates.io:andromeda-rates";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    msg.validate()?;
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    for strategy in msg.strategies {
+        STRATEGY_CONTRACT_ADDRESSES.save(
+            deps.storage,
+            strategy.strategy_type.to_string(),
+            &strategy.address,
+        )?;
+    }
+    ADOContract::default().instantiate(
+        deps.storage,
+        deps.api,
+        info,
+        BaseInstantiateMsg {
+            ado_type: "vault".to_string(),
+            operators: msg.operators,
+            modules: None,
+            primitive_contract: None,
+        },
+    )
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::AndrReceive(msg) => execute_andr_receive(deps, env, info, msg),
+        ExecuteMsg::Deposit {
+            recipient,
+            amount,
+            strategy,
+        } => execute_deposit(deps, env, info, amount, recipient, strategy),
+        ExecuteMsg::Withdraw {
+            recipient,
+            withdrawals,
+            strategy,
+        } => execute_withdraw(deps, env, info, recipient, withdrawals, strategy),
+    }
+}
+
+fn execute_andr_receive(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: AndromedaMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        AndromedaMsg::Receive(data) => {
+            let strategy: Option<StrategyType> = parse_message(&data)?;
+            let sender = info.sender.to_string();
+            execute_deposit(
+                deps,
+                env,
+                info,
+                None,
+                Some(Recipient::Addr(sender)),
+                strategy,
+            )
+        }
+        _ => ADOContract::default().execute(deps, env, info, msg, execute),
+    }
+}
+
+fn execute_deposit(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    amount: Option<Coin>,
+    recipient: Option<Recipient>,
+    strategy: Option<StrategyType>,
+) -> Result<Response, ContractError> {
+    require(
+        !info.funds.is_empty()
+            && (amount.is_none()
+                || (amount.is_some() && amount.clone().unwrap().amount.gt(&Uint128::zero()))),
```

Comment:
I think you can get rid of the `amount.is_some()` part since that line will only be evaluated when `amount.is_none()` is false. Also might be clearer to just use `!amount.clone.unwrap().amount.is_zero()`.

Reply:
Yep true, just artifact of an old conditional